### PR TITLE
disable Apple News pushes on non-production environments

### PR DESCRIPTION
**Pull request by mogmarsh**
## Summary

As titled.

## Notes for reviewers

None.

### CODE REVIEW

File: src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+use Alley\WP\Alleyvate\Feature;
+
+/**
+ * Disables Apple News Push on Non Production Environments.
+ */
+final class Disable_Apple_News_No_Prod_Push implements Feature {
+	/**
+	 * Store if this is a production environment.
+	 *
+	 * @var [type]
```
Comment (renatonascalves) [inline src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php:24]:
```suggestion
	 * @var bool
```

File: src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		add_filter( 'apple_news_skip_push', [ $this, 'filter_apple_news_skip_push' ], 1, 100 );
+		$this->is_production = $this->is_production_environment();
+	}
+
+	/**
+	 * Filter the Apple News push skip flag. If we are not on a production environment, skip the push.
+	 *
+	 * @param bool $skip Should we skip the Apple News push.
+	 */
+	public function filter_apple_news_skip_push( bool $skip ) {
```
Comment (renatonascalves) [inline src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php:41]:
```suggestion
	public function filter_apple_news_skip_push( bool $skip ): bool {
```

Comment (renatonascalves) [inline src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php:41]: Missing phpdoc for `@return`

File: src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		}
+		// If we are on Pantheon LIVE, return true.
+		if ( isset( $_ENV['PANTHEON_ENVIRONMENT'] ) && 'live' === $_ENV['PANTHEON_ENVIRONMENT'] ) {
+			return true;
+		}
+		// If we are on VIP Production, don't modify the value.
+		if ( \defined( 'VIP_GO_ENV' ) && VIP_GO_ENV === 'production' ) {
+			return true;
+		}
+		return false;
```
Comment (renatonascalves) [inline src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php:81]:
```suggestion
		return ( \defined( 'VIP_GO_ENV' ) && VIP_GO_ENV === 'production' );
```

**Review by renatonascalves — CHANGES_REQUESTED** [review]
Left a few questions and suggestions. :)

File: src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		/**
+		 * Filters the final value in filter_apple_news_skip_push.
+		 *
+		 * Allow developers to disable this feature.
+		 *
+		 * @since 2.0.0
+		 *
+		 * @param bool $skip Should we skip the Apple News push?
+		 */
+		return apply_filters( 'alleyvate_disable_apple_news_no_prod_push', true );
```
[the remark this answers is not in the document]
Comment (mogmarsh) [reply src/alley/wp/alleyvate/features/class-disable-apple-news-no-prod-push.php:57]: switching this to being filtered with `alleyvate_load_disable_apple_news_no_prod_push` in load.php

File: src/alley/wp/alleyvate/load.php [hunk 5]
```diff
@@ -27,6 +27,7 @@ function load(): void {
 	 * @var Feature[] $features
 	 */
 	$features = [
+		'disable_apple_new_push'        => new Features\Disable_Apple_News_No_Prod_Push(),
```
[the remark this answers is not in the document]
Comment (mogmarsh) [reply src/alley/wp/alleyvate/load.php:30]: went with `disable_apple_news_no_prod_push`

File: src/alley/wp/alleyvate/features/class-disable-apple-news-non-prod-push.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	 * @var bool
+	 */
+	private $is_production;
+
+	/**
+	 * Boot the feature.
+	 */
+	public function boot(): void {
+		add_filter( 'apple_news_skip_push', [ $this, 'filter_apple_news_skip_push' ], 1, 100 );
+		$this->is_production = $this->is_production_environment();
```
Comment (anubisthejackle) [inline src/alley/wp/alleyvate/features/class-disable-apple-news-non-prod-push.php:33]: It feels like the only purpose for this class parameter is to make it easier to test the class. If that's not the case, whats the other reason? If that is the case, I'd encourage you to reconsider. Testing is great, but when we have to jump through hoops to make the tests work, that speaks to a likely issue with our architecture.

File: src/alley/wp/alleyvate/features/class-disable-apple-news-non-prod-push.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		// If we are on Pantheon LIVE, return true.
+		if ( isset( $_ENV['PANTHEON_ENVIRONMENT'] ) && 'live' === $_ENV['PANTHEON_ENVIRONMENT'] ) {
+			return true;
+		}
+		// If we are on VIP Production, don't modify the value.
+		if ( \defined( 'VIP_GO_ENV' ) && VIP_GO_ENV === 'production' ) {
+			return true;
+		}
+		return false;
+	}
```
Comment (anubisthejackle) [inline src/alley/wp/alleyvate/features/class-disable-apple-news-non-prod-push.php:73]: This entire method feels like it has broader applicability than just this class. Could we move this to a helpers file? I think if we do that, we could then use it directly in the method where we're check `if ( $this->is_production ) {` instead, and then in the tests we can mock the helper function.

File: src/alley/wp/alleyvate/features/class-disable-apple-news-non-prod-push.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+		// All other cases, return true.
+
+		return true;
+	}
+
+	/**
+	 * Detect if we are on a production environment.
+	 *
+	 * Note: I thought about using https://developer.wordpress.org/reference/functions/wp_get_environment_type/
+	 * but it defaults to 'production' if no value is set, which did not seem ideal for this use case.
```
Comment (dlh01) [inline src/alley/wp/alleyvate/features/class-disable-apple-news-non-prod-push.php:55]: While a return value of `production` might be unreliable for this reason, is a return value other than `production` still reliable?

**Comment by jakewrfoster** [comment]
@mogmarsh does this close #26 ?

**Review by benpbolton — DISMISSED** [review]
Temporary review

**Comment by emilyatmobtown** [comment]
I'd recommend also disabling remote images when not on production. Otherwise, Apple News will trigger an error and fail the push if it tries to fetch remote images that it doesn't have access to, i.e. from local envs or protected non-production envs.

**Comment by mogmarsh** [comment]
> I'd recommend also disabling remote images when not on production. Otherwise, Apple News will trigger an error and fail the push if it tries to fetch remote images that it doesn't have access to, i.e. from local envs or protected non-production envs.

this is disabling all pushes, so the image references won't matter.

**Comment by emilyatmobtown** [comment]
@mogmarsh Ah, right. I was thinking of scenarios in which we have test credentials on non-prod. 

In that case, I'd suggest also disabling if `WP_CLI` is true. That will ensure that posts don't get pushed on batch updates via CLI in `production`.

**Comment by jakewrfoster** [comment]
@emilyatmobtown's input is valid, but should come in the form of a future issue and not block this one. I see preventing prod bulk CLI updates as a separate feature from blocking pushes on lower environments.

**Comment by mogmarsh** [comment]
> @emilyatmobtown's input is valid, but should come in the form of a future issue and not block this one. I see preventing prod bulk CLI updates as a separate feature from blocking pushes on lower environments.

https://github.com/alleyinteractive/wp-bulk-task/blob/main/src/trait-bulk-task-side-effects.php also could handle this in many CLI situations if we use it.

File: tests/bootstrap.php [hunk 9]
```diff
@@ -10,6 +10,13 @@
  * @package wp-alleyvate
  */
 
+/*
+ * This is required to override `wp_get_environment_type()`. Otherwise
+ * the method caches the initial result and always returns it, even if
+ * we modify the environment variable.
+ */
+\define( 'WP_RUN_CORE_TESTS', true ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedConstantFound
```
Comment (srtfisher) [inline tests/bootstrap.php:18]: fwiw, mantle will set this: https://github.com/alleyinteractive/mantle-framework/blob/4961ae56da6464d0b4732f075ea8e9ca2b3e2d89/src/mantle/testing/wordpress-bootstrap.php#L106-L108

**Review by srtfisher — APPROVED** [review]

File: tests/bootstrap.php (hunk 9 above)
Comment (anubisthejackle) [reply tests/bootstrap.php:18]: Interesting. It wasn't, which is why I had to add it here. Maybe worth a look?

Comment (anubisthejackle) [reply tests/bootstrap.php:18]: Ohhh I missed the `asof` line. That's probably why.

**Pull request merged**